source/val/validate_function.cpp

```diff
@@ -12,27 +12,102 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "source/val/validate.h"
-
 #include <algorithm>
 
 #include "source/opcode.h"
 #include "source/val/instruction.h"
+#include "source/val/validate.h"
 #include "source/val/validation_state.h"
 
 namespace spvtools {
 namespace val {
 namespace {
 
-// Returns true if |a| and |b| are instruction defining pointers that point to
-// the same type.
-bool ArePointersToSameType(val::Instruction* a, val::Instruction* b) {
+// Returns true if |a| and |b| are instructions defining types that are
+// isomorphic and the decorations that apply to |b| are a subset of the
+// decorations that apply to |a|.
+bool AreIsomorphicTypes(val::Instruction* a, val::Instruction* b,
```

Comment:
This is very similar to logically matching types for OpCopyLogical in SPIR-V 1.4. Take a look at `LogicallyMatch` in `validate_composites.cpp`. The SPIR-V spec also defines logically matching in OpCopyLogical's description.

Reply:
thanks that is very useful.  I have used it, but I had to change it a bit.  I also wanted to check decorations.